### PR TITLE
[PIWEB-14185] fix: Adapt detection of closing brackets in SearchConditionParser

### PR DESCRIPTION
**Pull request by czthiele**
- use smarter detection of closing bracket for condition values to avoid detecting brackets used as actual values
- add tests for SearchConditionParser

Per definition the closing bracket of a search condition is either followed by + for concatenation or by an end of string. We now match this using regex instead of simply searching for the next closing bracket. This way the parser can handle closing brackets inside the values, e.g. 123In[[1,2,3];[4,5,6],[7,8,9]] without breaking compatibility.

### CODE REVIEW

File: src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	using Zeiss.PiWeb.Api.Rest.Dtos.Data;
+
+	#endregion
+
+	[TestFixture]
+	public class SearchConditionParserTest
+	{
+		#region members
+
+		private static IEnumerable<string> _TestCases = new[]
```
Comment (KevinSchienerCZ) [inline src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs:28]: Please try to focus the tests on the edge cases, e.g. "123=[]", "123=[foo", "123=[fo]o", "123=[fo]o]", "123=[foo[]]", "123=[]foo]", ...

File: src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+			("123=[condition", false),
+			("123=condition]", false),
+			("123=[condit]ion", false),
+			("123=[condit]ion]", true),
+			("123=[condition[]]", true),
+			("123=[]condition]", true),
+			("123=[con[]dition]", true),
+			("123=[con[]dition]+124=[con[]dition]", true),
+			("123=[con[]dition]+124=[con[]dition", false)
+		};
```
Comment (MichaWresche) [inline src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs:43]: Since this test cases are only used by a single test method, can you annotate the test cases directly at the test method? This improves readability.

File: src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		[Test]
+		public void Parse_MissingOpeningBracket_ThrowsException()
+		{
+			Action action = () => SearchConditionParser.Parse( "123=value]" );
+			action.Should().Throw<InvalidOperationException>().Where( e => e.Message.Contains( "[]" ) );
+		}
+
+		[Test]
+		[TestCaseSource( nameof( _TestCases ) )]
+		public void Parse_CorrectSearchConditions_DoesNotThrow( (string, bool) testcase )
```
Comment (MichaWresche) [inline src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs:108]: The name implies that given test cases are correct but given test cases seems to can invalid. Can you adapt the test method?

**Review by MichaWresche — APPROVED** [review]
Minor test improvements

File: src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs (hunk 2 above)
Comment (czthiele) [reply src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs]: Done

File: src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs (hunk 3 above)
Comment (czthiele) [reply src/Api.Rest.Tests/Dtos/Data/SearchConditionParserTest.cs]: Done
